From 6919c644598a68f61c7c15852d467a64dd408f17 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Mon, 20 Oct 2025 16:56:30 +0100
Subject: [PATCH] fips: block HMAC calculation with unapproved digests

When using legacy HMAC APIs as used by older nodejs, dotnet, and
python - one can attempt to create HMAC using unapproved digest.

Block such access.
---
 crypto/hmac/hmac.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/crypto/hmac/hmac.c b/crypto/hmac/hmac.c
index 19fc7d3b4f..cb2923af4f 100644
--- a/crypto/hmac/hmac.c
+++ b/crypto/hmac/hmac.c
@@ -18,6 +18,7 @@
 #include <string.h>
 #include "internal/cryptlib.h"
 #include <openssl/opensslconf.h>
+#include <openssl/provider.h>
 #include <openssl/hmac.h>
 #include <openssl/core_names.h>
 #include "hmac_local.h"
@@ -30,6 +31,11 @@ int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int len,
     unsigned char pad[HMAC_MAX_MD_CBLOCK_SIZE];
     unsigned int keytmp_length;
     unsigned char keytmp[HMAC_MAX_MD_CBLOCK_SIZE];
+#ifndef FIPS_MODULE
+    const EVP_MD *checkmd = NULL;
+    const OSSL_PROVIDER *checkprovider = NULL;
+    const char *checkname = NULL;
+#endif
 
     /* If we are changing MD then we must have a key */
     if (md != NULL && md != ctx->md && (key == NULL || len < 0))
@@ -95,6 +101,23 @@ int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int len,
     }
     if (!EVP_MD_CTX_copy_ex(ctx->md_ctx, ctx->i_ctx))
         goto err;
+
+#ifndef FIPS_MODULE
+    /*
+     * Block non-fips provider digest in FIPS mode, note this has to
+     * be done this late, and using fetched MD from MD_CTX, as the
+     * input MD passed to this function can be NULL, or its provider
+     * might be NULL (unfetched). Also check all pointers, to be sure.
+     */
+    if (EVP_default_properties_is_fips_enabled(NULL) &&
+        (checkmd = EVP_MD_CTX_get0_md(ctx->md_ctx)) != NULL &&
+        (checkprovider = EVP_MD_get0_provider(checkmd)) != NULL &&
+        (checkname = OSSL_PROVIDER_get0_name(checkprovider)) != NULL &&
+        strcmp("fips", checkname)) {
+        return 0;
+    }
+#endif
+
     rv = 1;
  err:
     if (reset) {
-- 
2.51.0

